src/libsyntax/attr/builtin.rs

```diff
@@ -592,81 +592,87 @@ fn find_deprecation_generic<'a, I>(sess: &ParseSess,
     let diagnostic = &sess.span_diagnostic;
 
     'outer: for attr in attrs_iter {
-        if attr.path != "deprecated" {
-            continue
+        if !attr.check_name("deprecated") {
+            continue;
         }
 
-        mark_used(attr);
-
         if depr.is_some() {
             span_err!(diagnostic, item_sp, E0550, "multiple deprecated attributes");
             break
         }
 
-        depr = if let Some(metas) = attr.meta_item_list() {
-            let get = |meta: &MetaItem, item: &mut Option<Symbol>| {
-                if item.is_some() {
-                    handle_errors(sess, meta.span, AttrError::MultipleItem(meta.name()));
-                    return false
-                }
-                if let Some(v) = meta.value_str() {
-                    *item = Some(v);
-                    true
-                } else {
-                    if let Some(lit) = meta.name_value_literal() {
-                        handle_errors(
-                            sess,
-                            lit.span,
-                            AttrError::UnsupportedLiteral(
-                                "literal in `deprecated` \
-                                value must be a string",
-                                lit.node.is_bytestr()
-                            ),
-                        );
-                    } else {
-                        span_err!(diagnostic, meta.span, E0551, "incorrect meta item");
+        let meta = attr.meta().unwrap();
+        depr = match &meta.node {
+            MetaItemKind::Word => Some(Deprecation { since: None, note: None }),
+            MetaItemKind::NameValue(..) => {
+                Some(Deprecation {
+                    since: None,
+                    note: Some(meta.value_str().unwrap()),
```

Comment:
Well, this is a bit fragile since we are relying on the error recovery to be bad enough to stop compilation soon enough to avoid this panic for cases like `#[deprecated = 10]`.
That's why I recommended `map`ping the result of `value_str()` in https://github.com/rust-lang/rust/pull/58166#discussion_r253788490, it's more robust.

Reply:
It's more robust, but I think in a way that will mask the true cause of the issue. If the attribute parsing is relaxed in the future and lets an attribute like `#[deprecated = 10]` through, using a map will cause just the new rustdoc test to fail which could cause someone to think that the issue lies in rustdoc. With the unwrap, the `invalid-literal` test will ICE exactly where the new error handling needs to be.

Reply:
>If the attribute parsing is relaxed in the future and lets an attribute like `#[deprecated = 10]` through

I mean letting  `#[deprecated = 10]` through after the attribute validation error is reported, as an error recovery measure.
This is quite likely to happen in the future since deprecation checking needs to be moved to resolve time to treat imports correctly, and resolve runs even if there are attribute validation errors.